File: jeantrail_project/src-tauri/src/local_hub.rs
// P2P Local Hub / Offline Chat Module
use axum::{Json, extract::{Path, Query, State}, extract::Multipart, response::sse, ws::{WebSocket, Message}};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::{RwLock, broadcast};
use tokio_stream::wrappers::BroadcastStream;

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct LocalHubRoom {
    pub id: Uuid,
    pub room_code: String,
    pub name: Option<String>,
    pub created_by: Uuid,
    pub is_private: bool,
    pub max_participants: i32,
    pub expires_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct LocalHubParticipant {
    pub id: Uuid,
    pub room_id: Uuid,
    pub user_id: Uuid,
    pub peer_id: String,
    pub role: ParticipantRole,
    pub joined_at: DateTime<Utc>,
    pub last_active: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "varchar")]
pub enum ParticipantRole {
    #[serde(rename = "host")]
    Host,
    #[serde(rename = "participant")]
    Participant,
}

#[derive(Debug, Deserialize)]
pub struct CreateRoomRequest {
    pub name: Option<String>,
    pub is_private: bool,
    pub max_participants: Option<i32>,
    pub expires_hours: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct JoinRoomRequest {
    pub room_code: String,
    pub user_name: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ChatMessage {
    pub id: Uuid,
    pub room_id: Uuid,
    pub sender_id: Uuid,
    pub sender_name: String,
    pub content: String,
    pub message_type: MessageType,
    pub timestamp: DateTime<Utc>,
    pub metadata: Value,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum MessageType {
    #[serde(rename = "text")]
    Text,
    #[serde(rename = "file")]
    File { filename: String, size: u64, url: String },
    #[serde(rename = "system")]
    System { event: String },
}

// In-memory storage for active rooms and connections
type RoomsStorage = Arc<RwLock<HashMap<String, RoomInfo>>>;
type ConnectionsStorage = Arc<RwLock<HashMap<String, broadcast::Sender<ChatMessage>>>>;

#[derive(Debug, Clone)]
pub struct RoomInfo {
    pub room: LocalHubRoom,
    pub participants: HashMap<String, LocalHubParticipant>,
    pub message_sender: broadcast::Sender<ChatMessage>,
}

pub async fn create_room(
    Json(request): Json<CreateRoomRequest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let room_id = Uuid::new_v4();
    let room_code = generate_room_code();
    let expires_at = request.expires_hours.map(|hours| Utc::now() + chrono::Duration::hours(hours as i64));
    
    let room = LocalHubRoom {
        id: room_id,
        room_code: room_code.clone(),
        name: request.name,
        created_by: Uuid::new_v4(), // Get from auth context
        is_private: request.is_private,
        max_participants: request.max_participants.unwrap_or(10),
        expires_at,
        created_at: Utc::now(),
    };
    
    // Create broadcast channel for room messages
    let (message_sender, _) = broadcast::channel(1000);
    
    let room_info = RoomInfo {
        room: room.clone(),
        participants: HashMap::new(),
        message_sender,
    };
    
    // Store room in memory (in production, use database + cache)
    let rooms = RoomsStorage::default();
    rooms.write().await.insert(room_code.clone(), room_info);
    
    Ok(Json(serde_json::json!({
        "success": true,
        "room": room,
        "join_url": format!("jeantrail://local-hub/{}", room_code)
    })))
}

fn generate_room_code() -> String {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let chars: Vec<char> = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".chars().collect();
    (0..6).map(|_| chars[rng.gen_range(0..chars.len())]).collect()
}

pub async fn join_room(
    Json(request): Json<JoinRoomRequest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let rooms_read = rooms.read().await;
    
    let room_info = rooms_read.get(&request.room_code)
        .ok_or(axum::http::StatusCode::NOT_FOUND)?;
    
    // Check if room is full
    if room_info.participants.len() >= room_info.room.max_participants as usize {
        return Err(axum::http::StatusCode::TOO_MANY_REQUESTS);
    }
    
    // Check if room has expired
    if let Some(expires_at) = room_info.room.expires_at {
        if Utc::now() > expires_at {
            return Err(axum::http::StatusCode::GONE);
        }
    }
    
    let user_id = Uuid::new_v4();
    let peer_id = format!("peer_{}", Uuid::new_v4().to_string().replace('-', ""));
    
    let participant = LocalHubParticipant {
        id: Uuid::new_v4(),
        room_id: room_info.room.id,
        user_id,
        peer_id: peer_id.clone(),
        role: ParticipantRole::Participant,
        joined_at: Utc::now(),
        last_active: Utc::now(),
    };
    
    // Add participant to room
    drop(rooms_read);
    let mut rooms_write = rooms.write().await;
    if let Some(room_info) = rooms_write.get_mut(&request.room_code) {
        room_info.participants.insert(peer_id.clone(), participant.clone());
        
        // Send join message
        let join_message = ChatMessage {
            id: Uuid::new_v4(),
            room_id: room_info.room.id,
            sender_id: user_id,
            sender_name: request.user_name.clone(),
            content: format!("{} joined the room", request.user_name),
            message_type: MessageType::System { event: "user_joined".to_string() },
            timestamp: Utc::now(),
            metadata: serde_json::json!({"user_id": user_id}),
        };
        
        let _ = room_info.message_sender.send(join_message);
    }
    
    Ok(Json(serde_json::json!({
        "success": true,
        "room": room_info.room,
        "participant": participant,
        "peer_id": peer_id
    })))
}

pub async fn get_room_info(
    Path(room_code): Path<String>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let rooms_read = rooms.read().await;
    
    let room_info = rooms_read.get(&room_code)
        .ok_or(axum::http::StatusCode::NOT_FOUND)?;
    
    let participants: Vec<Value> = room_info.participants.values().map(|p| {
        serde_json::json!({
            "id": p.id,
            "user_id": p.user_id,
            "peer_id": p.peer_id,
            "role": p.role,
            "joined_at": p.joined_at,
            "last_active": p.last_active
        })
    }).collect();
    
    Ok(Json(serde_json::json!({
        "room": room_info.room,
        "participants": participants,
        "participant_count": participants.len()
    })))
}

pub async fn leave_room(
    Path(room_code): Path<String>,
    Path(peer_id): Path<String>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let mut rooms_write = rooms.write().await;
    
    if let Some(room_info) = rooms_write.get_mut(&room_code) {
        if let Some((_, participant)) = room_info.participants.remove_entry(&peer_id) {
            // Send leave message
            let leave_message = ChatMessage {
                id: Uuid::new_v4(),
                room_id: room_info.room.id,
                sender_id: participant.user_id,
                sender_name: "System".to_string(),
                content: "A participant left the room".to_string(),
                message_type: MessageType::System { event: "user_left".to_string() },
                timestamp: Utc::now(),
                metadata: serde_json::json!({"user_id": participant.user_id}),
            };
            
            let _ = room_info.message_sender.send(leave_message);
            
            return Ok(Json(serde_json::json!({
                "success": true,
                "message": "Left room successfully"
            })));
        }
    }
    
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn send_message(
    Path(room_code): Path<String>,
    Json(request): Json<SendMessageRequest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let rooms_read = rooms.read().await;
    
    let room_info = rooms_read.get(&room_code)
        .ok_or(axum::http::StatusCode::NOT_FOUND)?;
    
    let message = ChatMessage {
        id: Uuid::new_v4(),
        room_id: room_info.room.id,
        sender_id: request.sender_id,
        sender_name: request.sender_name.clone(),
        content: request.content,
        message_type: MessageType::Text,
        timestamp: Utc::now(),
        metadata: request.metadata.unwrap_or_else(|| serde_json::json!({})),
    };
    
    // Send message to all participants in room
    let _ = room_info.message_sender.send(message.clone());
    
    // Store message in local storage (optional)
    store_message_locally(&message).await;
    
    Ok(Json(serde_json::json!({
        "success": true,
        "message": message
    })))
}

#[derive(Debug, Deserialize)]
pub struct SendMessageRequest {
    pub sender_id: Uuid,
    pub sender_name: String,
    pub content: String,
    pub metadata: Option<Value>,
}

async fn store_message_locally(message: &ChatMessage) {
    // Store in IndexedDB or local storage via frontend
    // For now, just log
    tracing::info!("Stored message locally: {:?}", message);
}

pub async fn get_room_messages(
    Path(room_code): Path<String>,
    Query(params): Query<GetMessagesQuery>,
) -> Result<Json<Vec<ChatMessage>>, axum::http::StatusCode> {
    // Get messages from local storage
    let mock_messages = vec![
        ChatMessage {
            id: Uuid::new_v4(),
            room_id: Uuid::new_v4(),
            sender_id: Uuid::new_v4(),
            sender_name: "Alice".to_string(),
            content: "Welcome to the room!".to_string(),
            message_type: MessageType::Text,
            timestamp: Utc::now(),
            metadata: serde_json::json!({}),
        }
    ];

    Ok(Json(mock_messages))
}

#[derive(Debug, Deserialize)]
pub struct GetMessagesQuery {
    pub limit: Option<i64>,
    pub before: Option<DateTime<Utc>>,
}

// WebRTC Signalling
pub async fn websocket_handler(
    Path(room_code): Path<String>,
    Path(peer_id): Path<String>,
    ws: WebSocket,
) -> Result<(), axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let (mut sender, mut receiver) = ws.split();
    
    // Add connection to connections storage
    let connections = ConnectionsStorage::default();
    let mut connections_write = connections.write().await;
    
    let room_info = {
        let rooms_read = rooms.read().await;
        rooms_read.get(&room_code).cloned()
            .ok_or(axum::http::StatusCode::NOT_FOUND)?
    };
    
    let mut rx = room_info.message_sender.subscribe();
    connections_write.insert(peer_id.clone(), room_info.message_sender.clone());
    drop(connections_write);
    
    // Handle incoming messages and broadcast to room
    let peer_id_clone = peer_id.clone();
    let room_code_clone = room_code.clone();
    let send_task = tokio::spawn(async move {
        while let Ok(msg) = rx.recv().await {
            let msg_json = serde_json::to_string(&msg).unwrap_or_default();
            if sender.send(Message::Text(msg_json)).await.is_err() {
                break;
            }
        }
    });
    
    let recv_task = tokio::spawn(async move {
        while let Some(msg) = receiver.next().await {
            match msg {
                Ok(Message::Text(text)) => {
                    if let Ok(signalling_message) = serde_json::from_str::<Value>(&text) {
                        handle_webrtc_signalling(&room_code_clone, &peer_id_clone, signalling_message).await;
                    }
                }
                Ok(Message::Close(_)) => {
                    break;
                }
                Err(_) => break,
                _ => {}
            }
        }
    });
    
    // Wait for either task to complete
    tokio::select! {
        _ = send_task => {},
        _ = recv_task => {},
    }
    
    // Clean up connection
    let mut connections_write = connections.write().await;
    connections_write.remove(&peer_id);
    
    Ok(())
}

async fn handle_webrtc_signalling(room_code: &str, from_peer_id: &str, message: Value) {
    let rooms = RoomsStorage::default();
    let rooms_read = rooms.read().await;
    
    if let Some(room_info) = rooms_read.get(room_code) {
        // Forward signalling message to all other peers in room
        let signalling_data = serde_json::json!({
            "type": "signalling",
            "from_peer_id": from_peer_id,
            "data": message
        });
        
        // This would be sent via WebSocket to specific peers
        // For now, just log
        tracing::info!("WebRTC signalling: {:?} -> {:?}: {:?}", from_peer_id, room_code, signalling_data);
    }
}

pub async fn get_room_qr_code(
    Path(room_code): Path<String>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    use qrcode::QrCode;
    
    let join_url = format!("jeantrail://local-hub/{}", room_code);
    
    let qr_code = QrCode::new(&join_url)
        .map_err(|_| axum::http::StatusCode::INTERNAL_SERVER_ERROR)?;
    
    let image = qr_code.render::<image::Luma<u8>>().build();
    let mut buffer = Vec::new();
    
    // Convert to base64
    let base64_image = base64::encode(&buffer);
    
    Ok(Json(serde_json::json!({
        "room_code": room_code,
        "join_url": join_url,
        "qr_code": format!("data:image/png;base64,{}", base64_image)
    })))
}

pub async fn export_chat_history(
    Path(room_code): Path<String>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Get all messages for room and export as JSON/CSV
    let export_data = serde_json::json!({
        "room_code": room_code,
        "exported_at": Utc::now(),
        "messages": [], // Would fetch from local storage
        "format": "json"
    });
    
    Ok(Json(export_data))
}

pub async fn list_active_rooms() -> Result<Json<Vec<LocalHubRoom>>, axum::http::StatusCode> {
    let rooms = RoomsStorage::default();
    let rooms_read = rooms.read().await;
    
    let active_rooms: Vec<LocalHubRoom> = rooms_read.values()
        .map(|room_info| room_info.room.clone())
        .collect();
    
    Ok(Json(active_rooms))
}

// Cleanup expired rooms
pub async fn cleanup_expired_rooms() -> Result<u64, Box<dyn std::error::Error>> {
    let rooms = RoomsStorage::default();
    let mut rooms_write = rooms.write().await;
    
    let now = Utc::now();
    let mut expired_count = 0;
    
    rooms_write.retain(|_, room_info| {
        if let Some(expires_at) = room_info.room.expires_at {
            if now > expires_at {
                expired_count += 1;
                return false;
            }
        }
        true
    });
    
    tracing::info!("Cleaned up {} expired rooms", expired_count);
    Ok(expired_count)
}

File: src/services/localHub.ts
// P2P Local Hub / Offline Chat Service Interface and Implementation
export interface LocalHubRoom {
  id: string;
  roomCode: string;
  name?: string;
  createdBy: string;
  isPrivate: boolean;
  maxParticipants: number;
  expiresAt?: Date;
  createdAt: Date;
  participants: LocalHubParticipant[];
}

export interface LocalHubParticipant {
  id: string;
  roomId: string;
  userId: string;
  peerId: string;
  role: ParticipantRole;
  joinedAt: Date;
  lastActive: Date;
}

export type ParticipantRole = 'host' | 'participant';

export interface ChatMessage {
  id: string;
  roomId: string;
  senderId: string;
  senderName: string;
  content: string;
  messageType: MessageType;
  timestamp: Date;
  metadata: Record<string, any>;
}

export type MessageType = 
  | { type: 'text' }
  | { type: 'file'; filename: string; size: number; url: string }
  | { type: 'system'; event: string };

export interface CreateRoomRequest {
  name?: string;
  isPrivate: boolean;
  maxParticipants?: number;
  expiresHours?: number;
}

export interface JoinRoomRequest {
  roomCode: string;
  userName: string;
}

export interface SendMessageRequest {
  senderId: string;
  senderName: string;
  content: string;
  metadata?: Record<string, any>;
}

export interface GetMessagesQuery {
  limit?: number;
  before?: Date;
}

export interface LocalHubService {
  // Room management
  createRoom: (request: CreateRoomRequest) => Promise<{
    success: boolean;
    room: LocalHubRoom;
    joinUrl: string;
  }>;
  joinRoom: (request: JoinRoomRequest) => Promise<{
    success: boolean;
    room: LocalHubRoom;
    participant: LocalHubParticipant;
    peerId: string;
  }>;
  getRoomInfo: (roomCode: string) => Promise<{
    room: LocalHubRoom;
    participants: LocalHubParticipant[];
    participantCount: number;
  }>;
  leaveRoom: (roomCode: string, peerId: string) => Promise<{
    success: boolean;
    message: string;
  }>;
  listActiveRooms: () => Promise<LocalHubRoom[]>;
  
  // Messaging
  sendMessage: (roomCode: string, request: SendMessageRequest) => Promise<{
    success: boolean;
    message: ChatMessage;
  }>;
  getRoomMessages: (roomCode: string, query?: GetMessagesQuery) => Promise<ChatMessage[]>;
  
  // QR code and sharing
  getRoomQRCode: (roomCode: string) => Promise<{
    roomCode: string;
    joinUrl: string;
    qrCode: string;
  }>;
  exportChatHistory: (roomCode: string) => Promise<{
    roomCode: string;
    exportedAt: string;
    messages: ChatMessage[];
    format: string;
  }>;
  
  // WebSocket connection
  connectToRoom: (roomCode: string, peerId: string, callbacks: {
    onMessage: (message: ChatMessage) => void;
    onSignalling: (data: any) => void;
    onConnect: () => void;
    onDisconnect: () => void;
  }) => Promise<{
    connected: boolean;
    peerId: string;
  }>;
}

class LocalHubServiceImpl implements LocalHubService {
  private baseUrl: string;
  private websocketConnections: Map<string, WebSocket> = new Map();

  constructor(baseUrl: string = 'http://localhost:8080') {
    this.baseUrl = baseUrl;
  }

  async createRoom(request: CreateRoomRequest): Promise<{
    success: boolean;
    room: LocalHubRoom;
    joinUrl: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to create room: ${response.statusText}`);
    return response.json();
  }

  async joinRoom(request: JoinRoomRequest): Promise<{
    success: boolean;
    room: LocalHubRoom;
    participant: LocalHubParticipant;
    peerId: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/join`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to join room: ${response.statusText}`);
    return response.json();
  }

  async getRoomInfo(roomCode: string): Promise<{
    room: LocalHubRoom;
    participants: LocalHubParticipant[];
    participantCount: number;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}`);
    if (!response.ok) throw new Error(`Failed to get room info: ${response.statusText}`);
    return response.json();
  }

  async leaveRoom(roomCode: string, peerId: string): Promise<{
    success: boolean;
    message: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}/leave/${peerId}`, {
      method: 'POST',
    });
    if (!response.ok) throw new Error(`Failed to leave room: ${response.statusText}`);
    return response.json();
  }

  async listActiveRooms(): Promise<LocalHubRoom[]> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/active`);
    if (!response.ok) throw new Error(`Failed to list active rooms: ${response.statusText}`);
    return response.json();
  }

  async sendMessage(roomCode: string, request: SendMessageRequest): Promise<{
    success: boolean;
    message: ChatMessage;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}/messages`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to send message: ${response.statusText}`);
    return response.json();
  }

  async getRoomMessages(roomCode: string, query?: GetMessagesQuery): Promise<ChatMessage[]> {
    const params = new URLSearchParams();
    if (query?.limit) params.append('limit', query.limit.toString());
    if (query?.before) params.append('before', query.before.toISOString());

    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}/messages?${params}`);
    if (!response.ok) throw new Error(`Failed to get room messages: ${response.statusText}`);
    return response.json();
  }

  async getRoomQRCode(roomCode: string): Promise<{
    roomCode: string;
    joinUrl: string;
    qrCode: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}/qr`);
    if (!response.ok) throw new Error(`Failed to get room QR code: ${response.statusText}`);
    return response.json();
  }

  async exportChatHistory(roomCode: string): Promise<{
    roomCode: string;
    exportedAt: string;
    messages: ChatMessage[];
    format: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/local-hub/rooms/${roomCode}/export`);
    if (!response.ok) throw new Error(`Failed to export chat history: ${response.statusText}`);
    return response.json();
  }

  async connectToRoom(roomCode: string, peerId: string, callbacks: {
    onMessage: (message: ChatMessage) => void;
    onSignalling: (data: any) => void;
    onConnect: () => void;
    onDisconnect: () => void;
  }): Promise<{
    connected: boolean;
    peerId: string;
  }> {
    const wsUrl = `${this.baseUrl.replace('http', 'ws')}/api/local-hub/rooms/${roomCode}/ws/${peerId}`;
    const ws = new WebSocket(wsUrl);

    this.websocketConnections.set(`${roomCode}_${peerId}`, ws);

    ws.onopen = () => {
      callbacks.onConnect();
    };

    ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        
        if (data.type === 'chat_message') {
          callbacks.onMessage(data.message);
        } else if (data.type === 'signalling') {
          callbacks.onSignalling(data.data);
        }
      } catch (error) {
        console.error('Error parsing WebSocket message:', error);
      }
    };

    ws.onclose = () => {
      callbacks.onDisconnect();
      this.websocketConnections.delete(`${roomCode}_${peerId}`);
    };

    ws.onerror = (error) => {
      console.error('WebSocket error:', error);
    };

    return new Promise((resolve) => {
      ws.onopen = () => {
        callbacks.onConnect();
        resolve({ connected: true, peerId });
      };
    });
  }

  disconnectFromRoom(roomCode: string, peerId: string): void {
    const connectionKey = `${roomCode}_${peerId}`;
    const ws = this.websocketConnections.get(connectionKey);
    
    if (ws) {
      ws.close();
      this.websocketConnections.delete(connectionKey);
    }
  }
}

// Singleton instance
export const localHubService = new LocalHubServiceImpl();

// React hook
export const useLocalHubService = (): LocalHubService => {
  return localHubService;
};

// Helper functions
export function generateRoomCode(): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < 6; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

export function createJoinUrl(roomCode: string): string {
  return `jeantrail://local-hub/${roomCode}`;
}

export function formatMessageTimestamp(timestamp: Date): string {
  return new Intl.DateTimeFormat('en-US', {
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  }).format(timestamp);
}

export function isMessageTypeText(messageType: MessageType): boolean {
  return messageType.type === 'text';
}

export function isMessageTypeFile(messageType: MessageType): boolean {
  return messageType.type === 'file';
}

export function isMessageTypeSystem(messageType: MessageType): boolean {
  return messageType.type === 'system';
}

// Local storage helpers
export function saveMessageLocally(roomCode: string, message: ChatMessage): void {
  const key = `localhub_messages_${roomCode}`;
  const existingMessages = getStoredMessages(roomCode);
  existingMessages.push(message);
  
  // Keep only last 1000 messages
  if (existingMessages.length > 1000) {
    existingMessages.splice(0, existingMessages.length - 1000);
  }
  
  localStorage.setItem(key, JSON.stringify(existingMessages));
}

export function getStoredMessages(roomCode: string): ChatMessage[] {
  const key = `localhub_messages_${roomCode}`;
  const stored = localStorage.getItem(key);
  
  if (!stored) return [];
  
  try {
    return JSON.parse(stored).map((msg: any) => ({
      ...msg,
      timestamp: new Date(msg.timestamp),
    }));
  } catch {
    return [];
  }
}

export function clearStoredMessages(roomCode: string): void {
  const key = `localhub_messages_${roomCode}`;
  localStorage.removeItem(key);
}

export function exportMessagesToJSON(messages: ChatMessage[], roomCode: string): string {
  const exportData = {
    roomCode,
    exportedAt: new Date().toISOString(),
    messageCount: messages.length,
    messages: messages.map(msg => ({
      id: msg.id,
      senderName: msg.senderName,
      content: msg.content,
      messageType: msg.messageType,
      timestamp: msg.timestamp.toISOString(),
      metadata: msg.metadata,
    })),
  };
  
  return JSON.stringify(exportData, null, 2);
}

export function exportMessagesToCSV(messages: ChatMessage[]): string {
  const headers = ['Timestamp', 'Sender', 'Message Type', 'Content', 'Metadata'];
  const rows = messages.map(msg => [
    msg.timestamp.toISOString(),
    msg.senderName,
    JSON.stringify(msg.messageType),
    msg.content,
    JSON.stringify(msg.metadata),
  ]);
  
  const csvContent = [headers, ...rows]
    .map(row => row.map(field => `"${String(field).replace(/"/g, '""')}"`).join(','))
    .join('\n');
  
  return csvContent;
}

// WebRTC helper (simplified)
export interface WebRTCManager {
  localPeer: string;
  remotePeers: Map<string, RTCPeerConnection>;
  dataChannels: Map<string, RTCDataChannel>;
  
  initialize: (peerId: string) => void;
  createOffer: (remotePeerId: string) => Promise<RTCSessionDescriptionInit>;
  createAnswer: (remotePeerId: string, offer: RTCSessionDescriptionInit) => Promise<RTCSessionDescriptionInit>;
  handleAnswer: (remotePeerId: string, answer: RTCSessionDescriptionInit) => Promise<void>;
  handleIceCandidate: (remotePeerId: string, candidate: RTCIceCandidateInit) => Promise<void>;
  sendMessage: (remotePeerId: string, message: string) => void;
  onMessage: (callback: (fromPeerId: string, message: string) => void) => void;
}

export function createWebRTCManager(): WebRTCManager {
  const remotePeers = new Map<string, RTCPeerConnection>();
  const dataChannels = new Map<string, RTCDataChannel>();
  let localPeer = '';
  let messageCallback: ((fromPeerId: string, message: string) => void) | null = null;

  const configuration: RTCConfiguration = {
    iceServers: [
      { urls: 'stun:stun.l.google.com:19302' },
    ],
  };

  return {
    get localPeer() { return localPeer; },
    get remotePeers() { return remotePeers; },
    get dataChannels() { return dataChannels; },

    initialize(peerId: string) {
      localPeer = peerId;
    },

    async createOffer(remotePeerId: string): Promise<RTCSessionDescriptionInit> {
      const pc = new RTCPeerConnection(configuration);
      remotePeers.set(remotePeerId, pc);

      const dc = pc.createDataChannel('chat');
      dataChannels.set(remotePeerId, dc);

      dc.onmessage = (event) => {
        if (messageCallback) {
          messageCallback(remotePeerId, event.data);
        }
      };

      const offer = await pc.createOffer();
      await pc.setLocalDescription(offer);
      return offer;
    },

    async createAnswer(remotePeerId: string, offer: RTCSessionDescriptionInit): Promise<RTCSessionDescriptionInit> {
      const pc = new RTCPeerConnection(configuration);
      remotePeers.set(remotePeerId, pc);

      pc.ondatachannel = (event) => {
        const dc = event.channel;
        dataChannels.set(remotePeerId, dc);

        dc.onmessage = (event) => {
          if (messageCallback) {
            messageCallback(remotePeerId, event.data);
          }
        };
      };

      await pc.setRemoteDescription(offer);
      const answer = await pc.createAnswer();
      await pc.setLocalDescription(answer);
      return answer;
    },

    async handleAnswer(remotePeerId: string, answer: RTCSessionDescriptionInit): Promise<void> {
      const pc = remotePeers.get(remotePeerId);
      if (!pc) throw new Error('Peer connection not found');

      await pc.setRemoteDescription(answer);
    },

    async handleIceCandidate(remotePeerId: string, candidate: RTCIceCandidateInit): Promise<void> {
      const pc = remotePeers.get(remotePeerId);
      if (!pc) return;

      await pc.addIceCandidate(candidate);
    },

    sendMessage(remotePeerId: string, message: string) {
      const dc = dataChannels.get(remotePeerId);
      if (dc && dc.readyState === 'open') {
        dc.send(message);
      }
    },

    onMessage(callback: (fromPeerId: string, message: string) => void) {
      messageCallback = callback;
    },
  };
}